### PR TITLE
feat: Support for User Specifying Primary Key Scheme (default or cq-ids)

### PR DESCRIPTION
**Pull request by bbernays**
#### Summary

Initial implementation for https://github.com/cloudquery/cloudquery/issues/8811

---

Use the following steps to ensure your PR is ready to be reviewed

- [ ] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [ ] Run `go fmt` to format your code 🖊
- [ ] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [ ] Update or add tests 🧪
- [ ] Ensure the status checks below are successful ✅

### CODE REVIEW

**Review by erezrokah — COMMENTED** [review]
I know this is still in draft, but maybe since PKs and columns are defined on sources, this should be a source option?

I mean, what if I want `_cq_id` PK for AWS, and regular PKs for Azure?

**Comment by bbernays** [comment]
> I know this is still in draft, but maybe since PKs and columns are defined on sources, this should be a source option?
> 
> I mean, what if I want `_cq_id` PK for AWS, and regular PKs for Azure?

I am actually against putting it on the source because then the user will have to fetch the same data multiple times if they want to put the same data in multiple destinations (with different PK settings). In the case where a user has multiple sources and a single destination but want different behavior for each source, they should duplicate the destination.

**Comment by erezrokah** [comment]
> I am actually against putting it on the source because then the user will have to fetch the same data multiple times if they want to put the same data in multiple destinations (with different PK settings). In the case where a user has multiple sources and a single destination but want different behavior for each source, they should duplicate the destination.

Got it, makes sense

**Review by disq — APPROVED** [review]
LGTM, looked for a simpler approach but I don't see it

File: specs/pk_mode_test.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+		t.Fatalf("expected:\"cq-id\" got:%s", string(pkModeStr))
+	}
+
+	pkMode = PKModeCompositeKeys
+	if pkModeStr, err := pkMode.MarshalJSON(); err != nil {
+		t.Fatal(err)
+	} else if string(pkModeStr) != `"composite-keys"` {
+		t.Fatalf("expected:\"cq-id\" got:%s", string(pkModeStr))
+	}
+
```
Comment (erezrokah) [inline specs/pk_mode_test.go:49]:
```suggestion
```

File: schema/table.go [hunk 2]
```diff
@@ -314,6 +316,7 @@ func (t *Table) GetChanges(old *Table) []TableColumnChange {
 			})
 		}
 	}
+	// Remove columns that have been deleted from table definition.
```
Comment (erezrokah) [inline schema/table.go:319]:
```suggestion
	// A column was removed from the table definition
```

File: schema/table.go [hunk 3]
```diff
@@ -297,6 +297,7 @@ func (t *Table) GetChanges(old *Table) []TableColumnChange {
 	var changes []TableColumnChange
 	for _, c := range t.Columns {
 		otherColumn := old.Columns.Get(c.Name)
+		// Add in a column that didn't exist before.
```
Comment (erezrokah) [inline schema/table.go:300]:
```suggestion
		// A column was added to the table definition
```

File: schema/table.go [hunk 4]
```diff
@@ -305,6 +306,7 @@ func (t *Table) GetChanges(old *Table) []TableColumnChange {
 			})
 			continue
 		}
+		// column type changed or other column options changed (not null or PK)
```
Comment (erezrokah) [inline schema/table.go:309]:
```suggestion
		// Column type or options (e.g. PK, Not Null) changed in the new table definition
```

File: specs/pk_mode.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+type PKMode int
+
+const (
+	PKModeCompositeKeys PKMode = iota
+	PKModeCQID
+)
+
+var (
+	pkModeStrings = []string{"composite-keys", "cq-id"}
```
Comment (erezrokah) [inline specs/pk_mode.go:17]:
```suggestion
	pkModeStrings = []string{"composite-keys", "cq-id"}
```

I'm not sure `composite-keys` is accurate when there's a single PK.
Maybe call these `default` and `cq-id-only`?

**Review by erezrokah — APPROVED** [review]
Added a few comments (non blocking)

File: plugins/destination/plugin.go [hunk 6]
```diff
@@ -303,3 +306,20 @@ func setCqIDColumnOptionsForTables(tables []*schema.Table) {
...
+		for i, col := range table.Columns {
+			if col.Name == schema.CqIDColumn.Name {
+				table.Columns[i].CreationOptions.PrimaryKey = true
+				continue
+			}
+			if table.Parent != nil && col.Name == schema.CqParentIDColumn.Name {
+				table.Columns[i].CreationOptions.PrimaryKey = true
+				continue
+			}
+			table.Columns[i].CreationOptions.PrimaryKey = false
```
Comment (candiduslynx) [inline plugins/destination/plugin.go:321]:
```suggestion
			switch col.Name {
			case schema.CqIDColumn.Name:
				table.Columns[i].CreationOptions.PrimaryKey = true
			case schema.CqParentIDColumn.Name:
				table.Columns[i].CreationOptions.PrimaryKey = table.Parent != nil
			default:
				table.Columns[i].CreationOptions.PrimaryKey = false
			}
```

**Review by candiduslynx — APPROVED** [review]

**Review by yevgenypats — APPROVED** [review]

File: plugins/destination/plugin.go [hunk 7]
```diff
@@ -187,6 +187,9 @@ func (p *Plugin) Init(ctx context.Context, logger zerolog.Logger, spec specs.Des
 func (p *Plugin) Migrate(ctx context.Context, tables schema.Tables) error {
 	SetDestinationManagedCqColumns(tables)
 	setCqIDColumnOptionsForTables(tables)
+	if p.spec.PKMode == specs.PKModeCQID {
+		setCQIDAsPrimaryKeysForTables(tables)
```
Comment (yevgenypats) [inline plugins/destination/plugin.go:191]: Should that also be called on `Write`?

**Review by yevgenypats — CHANGES_REQUESTED** [review]
I think `setCQIDAsPrimaryKeysForTables` might need to be called also on destination `Write` . We have e2e test suites in the SDK under https://github.com/cloudquery/plugin-sdk/tree/main/internal/memdb we might want to see that this test is covered there as well.

File: plugins/destination/plugin_testing_migrate.go [hunk 8]
```diff
@@ -74,6 +74,17 @@ func testMigration(ctx context.Context, t *testing.T, p *Plugin, logger zerolog.
...
 
+	if p.spec.PKMode == specs.PKModeCQID {
+		for _, tColumn := range target.Columns {
+			if tColumn.Name != schema.CqIDColumn.Name {
+				continue
+			}
+			if !tColumn.CreationOptions.PrimaryKey {
+				return fmt.Errorf("expected primary key on %s", tColumn.Name)
+			}
+		}
```
Comment (candiduslynx) [inline plugins/destination/plugin_testing_migrate.go:85]:
```suggestion
		col := target.Get(schema.CqIDColumn.Name)
		if col == nil {
			return fmt.Errorf("expected to have %s column", schema.CqIDColumn.Name)
		}
		if !col.CreationOptions.PrimaryKey {
			return fmt.Errorf("expected primary key on %s", col.Name)
		}
```

**Review by candiduslynx — CHANGES_REQUESTED** [review]

File: plugins/destination/plugin_testing_migrate.go (hunk 8 above)
Comment (bbernays) [reply plugins/destination/plugin_testing_migrate.go]: Updated the test to fail if there is a PK on a different column than the cq_id

File: plugins/destination/plugin_testing_migrate.go [hunk 9]
```diff
@@ -74,6 +74,17 @@ func testMigration(ctx context.Context, t *testing.T, p *Plugin, logger zerolog.
...
 	}
 
+	if p.spec.PKMode == specs.PKModeCQID {
+		for _, tColumn := range target.Columns {
+			if tColumn.Name != schema.CqIDColumn.Name && tColumn.CreationOptions.PrimaryKey {
+				return fmt.Errorf("unexpected primary key on %s", tColumn.Name)
+			}
+			if tColumn.Name == schema.CqIDColumn.Name && !tColumn.CreationOptions.PrimaryKey {
+				return fmt.Errorf("expected primary key on %s", tColumn.Name)
+			}
```
Comment (candiduslynx) [inline plugins/destination/plugin_testing_migrate.go:84]:
```suggestion
			if tColumn.CreationOptions.PrimaryKey != tColumn.Name ==  schema.CqIDColumn.Name {
				return fmt.Errorf("primary key = %t for column %s", tColumn.CreationOptions.PrimaryKey, tColumn.Name)
			}
```

File: plugins/destination/plugin.go [hunk 10]
```diff
@@ -303,3 +305,24 @@ func setCqIDColumnOptionsForTables(tables []*schema.Table) {
...
+	}
+}
+func setCQIDAsPrimaryKeysForTables(tables schema.Tables) {
+	for _, table := range tables {
+		for i, col := range table.Columns {
+			switch col.Name {
+			case schema.CqIDColumn.Name:
+				table.Columns[i].CreationOptions.PrimaryKey = true
+			case schema.CqParentIDColumn.Name:
+				table.Columns[i].CreationOptions.PrimaryKey = table.Parent != nil
```
Comment (yevgenypats) [inline plugins/destination/plugin.go:321]: Why a parent should ever be a primary key ?

Comment (bbernays) [reply plugins/destination/plugin.go:321]: 2 different top level resources could resolve the same nested resource. In that case, the `cq_id` would be the same which means that one of the resources would be dropped

Comment (bbernays) [reply plugins/destination/plugin.go]: I have removed this because the `table.Parent` is always nil

File: plugins/destination/plugin.go [hunk 11]
```diff
@@ -303,3 +305,22 @@ func setCqIDColumnOptionsForTables(tables []*schema.Table) {
...
+}
+func setCQIDAsPrimaryKeysForTables(tables schema.Tables) {
+	for _, table := range tables {
+		for i, col := range table.Columns {
+			switch col.Name {
+			case schema.CqIDColumn.Name:
+				table.Columns[i].CreationOptions.PrimaryKey = true
+			default:
+				table.Columns[i].CreationOptions.PrimaryKey = false
+			}
```
Comment (candiduslynx) [inline plugins/destination/plugin.go:322]:
```suggestion
			table.Columns[i].CreationOptions.PrimaryKey = col.Name == schema.CqIDColumn.Name
```

**Review by candiduslynx — APPROVED** [review]
also a non-blocking nit

**Review by yevgenypats — APPROVED** [review]

**Pull request merged**